[video] Change Versions Art Handling

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Make the parent movie and versions behave more like sets and movies with respect to art:

- when browsing movies in the library, the art scraped/set by the user for the parent movie is used.
- choosing art from library manage context menu or info dialog sets the parent movie art.
- when browsing the versions of a movie (choose dialog from library, versions as a subfolder of the movie, versions top-level navigation), the art scraped/set by user for the version is used. If some types of art is missing at the version level, the art of the parent movie is used instead. Parent movie art will usually exist, since in v21 the movie is still the core element of the library.
- browsing versions/choosing art from Versions manager shows the art of the versions only, with no fallback on the art of the parent movie, for clarity.

Overall this allows different art for the group and each version, but if version art is not provided, group art will be used for everything.

Technically this is done by checking the path/options of the item at key points, and adding a special property to disable the art fallback that usually takes place, except when editing art.

This PR doesn't address the current issues of the Manage context menu and the Info dialog when launched for a version of a movie (from versions as subfolder of a movie or versions top-level navigation). The problems go beyond art. In that context, the dialog should probably replace some functions of Versions Manager and versions/extas buttons should be hidden. It may be v22 work though, an easier fix would be to remove access to the dialog.

Two small related changes included:
- [fix] ignore art with url field empty: setting art to "no art" doesn't delete the record in the art table but sets the url to blank. Those records have to be ignored for fallback between version and parent movie to work properly.
- [cosmetic] move some code of GetMoviesByWhere() into GetFilter() where it belongs

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Address some surprising behavior of art in master that are consequences of always using version art first, even when browsing movies in the library.
Attempt to provide a no-change solution for the most common case of single version for movies and small code change when there are multiple versions.

From the user point of view should feel similar to what will be available later in v22, although the implementation is not clean due to constraints of the current design.

This superseeds my previous attempts.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Movies with single and multiple versions.
Changed art with context menu > choose art, with info dialog Choose art
Changed art with Versions Manage dialog.
Browsed the library by titles, from versions navigation, in the versions subfolders.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Intuitive behavior of art.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[video] use version specific art only when navigating individual versions. Movie art otherwise.

CrystalP committed:
[video] Show version art with fallback to movie art, except when choosing version art

CrystalP committed:
[videodb] Ignore empty art

url is set to empty for Choose Art > no art

CrystalP committed:
[videodb] Move default version filter into GetFilter()

ksooo commented on xbmc/video/dialogs/GUIDialogVideoInfo.cpp:
```diff
@@ -847,7 +847,7 @@ void AddCurrentArtTypes(std::vector<std::string>& artTypes, const CVideoInfoTag&
 
   if (tag.GetAssetInfo().GetId() >= 0)
-    db.GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, currentArt);
+    db.GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, false, currentArt);
```
You know I do not like bool parameters in public interfaces. ;-) What does this new `false` tell me? I need to look up the method declaration in the respective header file to get to know. If you would use an enum class with speaking values, understanding this code would be much easier. Any change ti change the bool param to an enum class param?

CrystalP replied:
I saw this coming but am still thinking about a better way. Likely will be a different overload of GetArtForAsset with less parameters when fallback is not wanted. Inside videodb it would still be the same.

CrystalP replied:
and some doxy...

ksooo replied:
> Likely will be a different overload of GetArtForAsset with less parameters

That would be even more elegant, yes.

CrystalP replied:
the overload turned out cumbersome, then I simplified the function, in the end it's an enum.

ksooo commented on xbmc/video/VideoThumbLoader.cpp:
```diff
@@ -397,7 +397,8 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
     m_videoDatabase->Open();
 
-    if (tag.GetAssetInfo().GetId() >= 0)
+    const CURL itemUrl{item.GetPath()};
+    if (itemUrl.HasOption("videoversionid"))
```
Why this change? Just want to understand...

CrystalP replied:
This is how the code knows that we're browsing movies in the library and considering the movies as groups of versions (use "movie" art type) or using one of the paths that displays the individual versions (use "videoversion" art type): versions in folder, choose version dialog, versions navigation, versions manager dialog.

Same could be done with a GetPath() and looking for a videodb://movies/videoversions/ prefix.

It's not forward looking but works today. In v22 this will have to be redone completely, stuff moved around, ...

ksooo replied:
Okay, thanks for the explanation.

CrystalP replied:
refactored into a method of CFileItem as the same condition will have to be used to fix stuff in info dialog / context menu opened on versions, if the concept of this PR is accepted.

ksooo commented on xbmc/video/VideoThumbLoader.cpp:
```diff
@@ -400,7 +400,12 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
     {
-      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, artwork))
+      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type,
+                                          item.HasProperty("noartfallbacktoowner") &&
+                                                  item.GetProperty("noartfallbacktoowner") == true
+                                              ? false
+                                              : true,
```
Not 100% sure, but imo the whole term could be shortened to `!item.GetProperty("noartfallbacktoowner").asBoolen(false)`.

CrystalP replied:
yes. that's convenient!

CrystalP committed:
[videodb] Simplify GetArtForAsset()

The asset owner can be retrieved from the db and doesn't have to be provided through parameters.

CrystalP committed:
[videodb] optimize and make GetArtForAsset() more readable

WHERE ( x, y) IN () is a non-standard SQL extension supported by sqlite, mySql and MariaDB

ksooo commented on xbmc/FileItem.cpp:
```diff
@@ -4273,3 +4273,14 @@ bool CFileItem::HasVideoExtras() const
+{
+  // @todo maybe in the future look for prefix videodb://movies/videoversions in path instead
+  // @todo better encoding of video assets as path, they won't always be tied with movies.
+  const CURL itemUrl{GetPath()};
```
Maybe add a check for being a videodb item first? `if (IsVideoDb()) ...` This is cheap and will avoid parsing of the URL etc. in many cases.

ksooo commented on xbmc/video/VideoDatabase.h:
```diff
@@ -401,6 +401,12 @@ const struct SDbTableOffsets DbMusicVideoOffsets[] =
   { VIDEODB_TYPE_INT, my_offsetof(CVideoInfoTag,m_iIdUniqueID)}
 };
 
+typedef enum
```
I would suggest an `enum class`. In general we want to get rid of classic enum's and typedef's if possible, because enum classes have some benefits, for example you can forward declare them.

ksooo commented on xbmc/video/VideoThumbLoader.cpp:
```diff
@@ -397,9 +397,14 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
     {
-      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, artwork))
+      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId,
+                                          item.GetProperty("noartfallbacktoowner").asBoolean(false)
+                                              ? ARTFALLBACK_NONE
+                                              : ARTFALLBACK_PARENT,
```
I like. 👍

CrystalP commented:
Thanks for the suggestions, applied!

CrystalP commented:
Jenkins build this please

ksooo reviewed (approved):
LGTM
